common/src/main/java/io/netty/util/internal/PlatformDependent.java

```diff
@@ -1018,51 +1019,48 @@ private static boolean isRoot0() {
         return false;
     }
 
-    @SuppressWarnings("LoopStatementThatDoesntLoop")
     private static int javaVersion0() {
-        int javaVersion;
+        final int majorVersion;
 
-        // Not really a loop
-        for (;;) {
-            // Android
-            if (isAndroid()) {
-                javaVersion = 6;
-                break;
-            }
+        if (isAndroid()) {
+            majorVersion = 6;
+        } else {
+            majorVersion = majorVersionFromJavaSpecificationVersion();
+        }
 
-            try {
-                Method getVersion = java.lang.Runtime.class.getMethod("version");
-                Object version = getVersion.invoke(null);
-                javaVersion = (Integer) version.getClass().getMethod("major").invoke(version);
-                break;
-            } catch (Throwable ignored) {
-                // Ignore
-            }
+        logger.debug("Java version: {}", majorVersion);
 
-            try {
-                Class.forName("java.time.Clock", false, getClassLoader(Object.class));
-                javaVersion = 8;
-                break;
-            } catch (Throwable ignored) {
-                // Ignore
-            }
+        return majorVersion;
+    }
 
-            try {
-                Class.forName("java.util.concurrent.LinkedTransferQueue", false, getClassLoader(BlockingQueue.class));
-                javaVersion = 7;
-                break;
-            } catch (Throwable ignored) {
-                // Ignore
-            }
+    static int majorVersionFromJavaSpecificationVersion() {
+        try {
+            final String javaSpecVersion = AccessController.doPrivileged(new PrivilegedAction<String>() {
+                @Override
+                public String run() {
+                    return System.getProperty("java.specification.version");
+                }
+            });
+            return majorVersion(javaSpecVersion);
+        } catch (SecurityException e) {
```

Comment:
In general, catching and swallowing the top-level `Throwable` is something that I think should be avoided. This means also catching `Error` for which there's generally no expectation that there is a clean recovery (e.g., `OutOfMemoryError`). This is especially the case for a foundational library like Netty; catching and swallowing `Throwable` hides these unrecoverable `Error`s from the application layer leaving them running on JVMs that are in a questionable state after such `Error`s.

It also hides bugs. In this case, string parsing can only fail if there is a bug in the parsing code or someone is running on a JVM that violates the `java.specification.version` format:

> Specification version numbers use a Dewey decimal notation consisting of numbers seperated by periods.

And from JEP-223:

>A version number, $VNUM, is a non-empty sequence of elements separated by period characters (U+002E). An element is either zero, or an unsigned integer numeral without leading zeros. The final element in a version number must not be zero. The format is:

> `^[1-9][0-9]*(((\.0)*\.[1-9][0-9]*)*)*$`

Neither of these should be hidden, they should bubble up with the fury of a lava bubble so that it can be addressed (by fixing the broken code, or moving to a non-broken JVM).

In elastic/elasticsearch#19231, we've gone to great lengths to eliminate catching `Throwable` from the codebase (to set the stage for elastic/elasticsearch#19272), and if I had my druthers I'd remove it from *every* library that Elasticsearch depends on because we are still susceptible to swallowed `OutOfMemoryError` and other such `Error`s.